### PR TITLE
Minimize quad trees requests for prefetch

### PR DESCRIPTION
**Pull request by LiubovDidkivska**
While prefetching quad trees we need to slice
tiles for requests from max level to min level
with predefined depth 4. In case if min level
allows to go up, we will request quad tree on
minimum possible level with depth 4. When min
level is to low we will start from 0 level, but
should continue slicing considering max level
and go up.
Changes related to not changing max level,
that we always have minimum requests, even if
some quads overlap.

Relates-To: OLPEDGE-2225
Signed-off-by: Liubov Didkivska <ext-liubov.didkivska@here.com>

### CODE REVIEW

File: olp-cpp-sdk-dataservice-read/tests/PrefetchRepositoryTest.cpp [hunk 1]
```diff
@@ -239,4 +240,58 @@ TEST(PrefetchRepositoryTest, GetSlicedTilesSiblings) {
   ASSERT_EQ(root_tiles_depth.begin()->first, parent1);
   ASSERT_EQ(root_tiles_depth.begin()->second, 4);
 }
+
+TEST(PrefetchRepositoryTest, GetSlicedTilesLowLevels) {
+  auto tile1 = olp::geo::TileKey::FromHereTile("23618366");
```
Comment (kzvieriev) [inline olp-cpp-sdk-dataservice-read/tests/PrefetchRepositoryTest.cpp:245]:
```suggestion
  auto tile = olp::geo::TileKey::FromHereTile("23618366");
```
There is no tile2 in the scope

File: olp-cpp-sdk-dataservice-read/src/repositories/PrefetchTilesRepository.cpp [hunk 2]
```diff
@@ -126,10 +126,10 @@ RootTilesForRequest PrefetchTilesRepository::GetSlicedTiles(
       if (min_level > levels_up) {
         min_level = min_level - levels_up;
       } else {
-        // if min_level is less than steps we need to go up, go some steps down
-        auto levels_down = levels_up - min_level;
+        // if min_level is less than steps we need to go up, set min level to
+        // zero. Some quads will overlap, but we will minimize quad tree
+        // requests
         min_level = 0u;
```
Comment (kzvieriev) [inline olp-cpp-sdk-dataservice-read/src/repositories/PrefetchTilesRepository.cpp:132]: I wonder, can overlapping quad trees break our protect/release mechanism?
e.g. 2 protected quads share 1 protected tile, after releasing the tile second quad will be still protected.

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.cpp [hunk 3]
```diff
@@ -47,55 +47,58 @@ const cache::KeyValueCache::KeyListType&
...
-        // no more protected tiles associated with this quad tree
-        // can add key for quad tree to be released and remove from map
-        keys_to_release_.emplace_back(
-            partitions_cache_repository_.CreateQuadKey(
-                layer_id_, it->first, kQuadTreeDepth, version_));
-        quad_trees_with_protected_tiles_.erase(it);
+void ReleaseDependencyResolver::ProcessTileKey(const geo::TileKey& tile_key) {
+  bool add_key = true;
+  auto process_tile = [&](const geo::TileKey& quad_root,
+                          const geo::TileKey& tile_key) {
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.cpp:58]: Might be an issue with a duplicate name for tile_key. I would find another name, e.g. tile

File: olp-cpp-sdk-dataservice-read/tests/PrefetchRepositoryTest.cpp [hunk 4]
```diff
@@ -239,4 +240,58 @@ TEST(PrefetchRepositoryTest, GetSlicedTilesSiblings) {
...
+  {
+    SCOPED_TRACE(
+        "Get sliced tiles with min/max levels less than requested tiles "
+        "levels");
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/tests/PrefetchRepositoryTest.cpp:249]: Please add a smaller trace.

File: olp-cpp-sdk-dataservice-read/src/repositories/PrefetchTilesRepository.cpp (hunk 2 above)
Comment (LiubovDidkivska) [reply olp-cpp-sdk-dataservice-read/src/repositories/PrefetchTilesRepository.cpp:132]: Release changed to consider all possible quads to releas

File: tests/integration/olp-cpp-sdk-dataservice-read/VersionedLayerClientTest.cpp [hunk 5]
```diff
@@ -4086,6 +4085,100 @@ TEST_F(DataserviceReadVersionedLayerClientTest, CatalogEndpointProvider) {
...
+                             .WithTileKeys({tile_key})
+                             .WithMinLevel(11)
+                             .WithMaxLevel(15);
+    auto future = client.PrefetchTiles(request).GetFuture();
+
+    ASSERT_NE(future.wait_for(kWaitTimeout), std::future_status::timeout);
+    ASSERT_TRUE(future.get().IsSuccessful());
+  }
+
+  ASSERT_TRUE(settings_.cache->Protect({cache_quad_key_5901734}));
```
Comment (kzvieriev) [inline tests/integration/olp-cpp-sdk-dataservice-read/VersionedLayerClientTest.cpp:4167]: Move check to the test case scope, please :-)

**Pull request merged**